lexint.js

```diff
@@ -0,0 +1,105 @@
+module.exports = {
+  preencode,
+  encode,
+  decode
+}
+
+function preencode (state, num) {
+  if (num < 251) {
+    state.end++
+  } else if (num < 256) {
+    state.end += 2
+  } else if (num < 0x10000) {
+    state.end += 3
+  } else if (num < 0x1000000) {
+    state.end += 4
+  } else if (num < 0x100000000) {
+    state.end += 5
+  } else {
+    state.end++
+    const exp = Math.floor(Math.log(num) / Math.log(2)) - 32
+    preencode(state, exp)
+    state.end += 6
+  }
+}
+
+function encode (state, num) {
+  const max = 251
+  const x = num - max
+
+  if (num < max) {
+    state.buffer[state.start++] = num
+  } else if (num < 256) {
+    state.buffer[state.start++] = max
+    state.buffer[state.start++] = x
+  } else if (num < 0x10000) {
+    state.buffer[state.start++] = max + 1
+    state.buffer[state.start++] = x >> 8 & 0xff
+    state.buffer[state.start++] = x & 0xff
+  } else if (num < 0x1000000) {
+    state.buffer[state.start++] = max + 2
+    state.buffer[state.start++] = x >> 16
+    state.buffer[state.start++] = x >> 8 & 0xff
+    state.buffer[state.start++] = x & 0xff
+  } else if (num < 0x100000000) {
+    state.buffer[state.start++] = max + 3
+    state.buffer[state.start++] = x >> 24
+    state.buffer[state.start++] = x >> 16 & 0xff
+    state.buffer[state.start++] = x >> 8 & 0xff
+    state.buffer[state.start++] = x & 0xff
+  } else {
+    // need to use Math here as bitwise ops are 32 bit
+    const exp = Math.floor(Math.log(x) / Math.log(2)) - 32
+    state.buffer[state.start++] = 0xff
+
+    encode(state, exp)
+    const rem = x / Math.pow(2, exp - 11)
+
+    for (let i = 5; i >= 0; i--) {
+      state.buffer[state.start++] = rem / Math.pow(2, 8 * i) & 0xff
+    }
+  }
+}
+
+function decode (state) {
+  const max = 251
+  const flag = state.buffer[state.start++]
+
+  if (flag < max) return flag
+
+  if (flag < 252) {
+    return state.buffer[state.start++] +
+      max
+  }
+
+  if (flag < 253) {
+    return (state.buffer[state.start++] << 8) +
+      state.buffer[state.start++] +
+      max
+  }
+
+  if (flag < 254) {
+    return (state.buffer[state.start++] << 16) +
+      (state.buffer[state.start++] << 8) +
+      state.buffer[state.start++] +
+      max
+  }
+
+  // << 24 result may be interpreted as negative
+  if (flag < 255) {
```

Comment:
`if (state.end - state.start < 4) throw new Error('Out of bounds')`

Same for the other branches in here :)

Reply:
https://github.com/compact-encoding/compact-encoding/blob/e58f23b17bc0c1a76960889889dc2493a48bb0fc/lexint.js#L71-L73